src/Client.java

```diff
@@ -123,16 +101,78 @@ private static void startGame(BufferedInputStream input, int[][] board) throws I
+    private static void readMove(BufferedOutputStream output, BufferedReader console, int[][] board, boolean joueurRouge) throws IOException {
+        //String move = console.readLine();
+
+        Map<String, List<int[]>> hashMove = getHashafficherMovesPossibles(board, joueurRouge);
+        Random random = new Random();
+        List<String> keysAsArray = new ArrayList<>(hashMove.keySet());
+
+        int[] randomMove = null;
+        String randomKey = null;
+        List<int[]> movesList = null;
+
+        if (!keysAsArray.isEmpty()) {
+            while (movesList == null || movesList.isEmpty()) {
+                // Pick a random key from the keysAsArray.
+                randomKey = keysAsArray.get(random.nextInt(keysAsArray.size()));
+                movesList = hashMove.get(randomKey);
+            }
+            randomMove = movesList.get(random.nextInt(movesList.size()));
+        }
+
+        System.out.println("Random Key: " + randomKey);
+        System.out.println("Random Value: " + Arrays.toString(randomMove));
+
+        String [] old_coords = randomKey.split(" ");
+        old_coords[0] = old_coords[0].substring(1);
+        old_coords[1] = old_coords[1].substring(1);
+
+        String move = new Move(Integer.parseInt(old_coords[0]), Integer.parseInt(old_coords[1]), randomMove[0], randomMove[1]).s; //Structure si on veut convertir coordonnées en String demandé par serveur
+
+        updateBoard(move, board);
         output.write(move.getBytes(), 0, move.length());
         output.flush();
     }
 
+    public static Map<String, List<int[]>> getHashafficherMovesPossibles(int[][] board, boolean joueurRouge) {
+        Map<String, List<int[]>> allPossibleMoves = getAllPossibleMoves(board, joueurRouge);
+        for (Map.Entry<String, List<int[]>> entry : allPossibleMoves.entrySet()) {
+            System.out.print(entry.getKey() + " : [");
+            for (int[] coordinates : entry.getValue()) {
+                System.out.print(Arrays.toString(coordinates) + ", ");
+            }
+            System.out.println("]");
+        }
+        return allPossibleMoves;
+    }
+    public static Map<String, List<int[]>> getAllPossibleMoves(int[][] board, boolean joueurRouge) {
+        Map<String, List<int[]>> allPossibleMoves = new HashMap<>();
+
+        for (int x = 0; x < board.length; x++) {
+            for (int y = 0; y < board[x].length; y++) {
+                if (board[x][y] == 4 && joueurRouge) {
+                    List<int[]> possibleMoves = getPossibleMoves(board, x, y);
+
+                    String position = "x" + x + " y" + y;
+
+                    allPossibleMoves.put(position, possibleMoves);
+                }else if((board[x][y] == 5 || board[x][y] == 2) && !joueurRouge){
+                    List<int[]> possibleMoves = getPossibleMoves(board, x, y);
+
+                    String position = "x" + x + " y" + y;
+
+                    allPossibleMoves.put(position, possibleMoves);
+                }
```

Comment:
Le code est dupliqué. il serait peut-être plus simple de le faire ainsi:

```suggestion
                if ((board[x][y] == 4 && joueurRouge) || (board[x][y] == 5 || board[x][y] == 2) && !joueurRouge)) {
                    List<int[]> possibleMoves = getPossibleMoves(board, x, y);
                    String position = "x" + x + " y" + y;

                    allPossibleMoves.put(position, possibleMoves);
             }
```

Reply:
Ah oui bien vu. Faudrait pas rajouter une paranthèse : ((board[x][y] == 4 && joueurRouge) || **(**(board[x][y] == 5 || board[x][y] == 2) && !joueurRouge)**)**) ou ça revient au même ?

Reply:
J'ai modifié le if donc ca devrait être bon maintenant. Il manquait bien une paranthèse mb. :D